records/2024-01-22.reshaping-cicd-environments.md

```diff
@@ -0,0 +1,127 @@
+
+### More automation
+
+More automation for the release process is desired for the team to lessen overall
+burden of our support role. These are identified areas for automation:
+
+- Release changes approval,
+- Components (API, Admin, DD-API, documentation) releases to production,
+- Lambdas releases to staging,
+- Lambdas releases to production,
+- Tests reports are published for the team to see.
+
+### Lighter release process
+
+The team desire less chores for the ops lead and incident commander, to make
+room for other tasks that cannot benefit of automation.
+
+### More release autonomy
+
+Delegate release checks to developers in order to remove burden on the ops
+lead. By having developers approve and promote their own changes ready for a
+production release, this would remove some hurdles.
+
+### Safe release to production
+
+The releases need to happen with a certain safety of checks, verifications and
+certified approvals by the change owners. Human interventions might be required
+at certain points to approve of changes or identify issues, potentially
+reverting bad rollouts. At the same time, having too many human intervention
+will lead to a higher number of mistakes.
+
+### Isolated environments for testing
+
+There are changes of certain nature such as infrastructure that leads to
+disruption of the environment itself and might require isolation in a separate
+environment.
+
+## Options
+
+This section lists suggestions made by the team with different considerations
+and goals. We'll first review the current CI/CD process.
+
+### Current process
+
+![Current CI/CD workflow](https://raw.githubusercontent.com/cds-snc/notification-adr/main/records/diagrams/current-cicd.drawio.png?raw=true "Current CI/CD workflow")
+
+### Stronger responsibility for release changes on developers
+
+A lightweight changes that the process can include in the first place is to put
+more burden on the developers' shoulders on the changes approval step.
+Specifically, whatever the developer pushes to the staging environment is what
+will get released by the ops lead.
+
+This keeps the current environments as-is and does not make new automation
+additions. This is a status quo for most release di mensions except for the
+manual aspect.
+
+#### Pros
+
+- Incremental
+- Very cheap to implement
+- Lighten load on the ops lead role
+
+#### Cons
+
+- Higher risk factor of deploying to production (as it relies on multiple human interventions)
+
+### Create dedicated "Core" and "Dev" environments
+
+Currently the staging environment is used as an integration/development environment. This has been workng, but is not aligned with best practices, as it leaves staging in a non-production like state. With the introduction of the new "Dev" environment, it became possible to add an additional environment into the process that allows staging to always remain in the same state as production. 
```

Comment:
Can we add when it hasn't been working? Right now the only thing we test on staging is callbacks I believe.

Reply:
I am not sure if I understand correctly the question with reference to the callbacks. Can you expand on that? or hit me on Slack?